google/cloud/ndb/polymodel.py

```diff
@@ -184,17 +184,14 @@ class PolyModel(model.Model):
 
     class_ = _ClassKeyProperty()
 
-    _class_map = {}  # Map class key -> suitable subclass.
-
     @classmethod
     def _update_kind_map(cls):
         """Override; called by Model._fix_up_properties().
 
         Update the kind map as well as the class map, except for PolyModel
         itself (its class key is empty).  Note that the kind map will
         contain entries for all classes in a PolyModel hierarchy; they all
-        have the same kind, but different class names.  PolyModel class
-        names, like regular Model class names, must be globally unique.
+        have the same kind, but different class names.
         """
         cls._kind_map[cls._class_name()] = cls
```

Comment:
This is the reason why class names have to be globally unique, unfortunately. Different model classes will overwrite each other here if they have the same class name.

You can see this by adding the following to the end of your test below:

```python
assert isinstance(CatCommand._kind_map["Cat"].bar, model.StringProperty)
assert not isinstance(CatAnimal._kind_map["Cat"].bar, model.StringProperty)
```

Notice how on the second line, even though we're accessing it through `CatAnimal`, the Model returned by `CatAnimal._kind_map["Cat"]` is a `CatCommand`.

Reply:
Yes, they are overwritten, but there's a code above it that uses its own dictionary using fully qualified names to look up for the real entity during model instance creation. So, yes, non-unique polymorphic classes, when accessed using `_kind_map` will be buggy, but why would a polymodel be looked up via global registry of top level classes? Polymodels provide their own registry and don't care what other registries think of them.
At least my legacy code with my fix doesn't experience any problems in my scenario as it did not in legacy ndb. And I use polymorphic models actively.

Reply:
Looks like I can override `_class_name` for `PolyModel` to return something like `Animal.Cat` and `Command.Cat`. In this case even `_kind_map` will keep its invariant.